dbcon/execplan/rewrites.cpp

```diff
@@ -0,0 +1,417 @@
+{
+  if (stackframe.direction == GoTo::Right)
+    stackframe.containsLeft = containsflag;
+  else
+    stackframe.containsRight = containsflag;
+}
+
+void fixUpTree(execplan::ParseTree** node, ChildType ltype, ChildType rtype,
+               StackFrame* parentframe = nullptr)
+{
+  if (ltype == ChildType::Leave)
+  {
+    if (rtype != ChildType::Leave)
+    {
+      execplan::ParseTree* oldNode = *node;
+      if (rtype == ChildType::Delete)
+        deleteOneNode((*node)->rightRef());
+      *node = (*node)->left();
+      deleteOneNode(&oldNode);
+    }
+  }
+  else
+  {
+    if (ltype == ChildType::Delete)
+      deleteOneNode((*node)->leftRef());
+    if (rtype == ChildType::Leave)
+    {
+      execplan::ParseTree* oldNode = *node;
+      *node = (*node)->right();
+      deleteOneNode(&oldNode);
+    }
+    else
+    {
+      if (rtype == ChildType::Delete)
+        deleteOneNode((*node)->rightRef());
+
+      if (parentframe)
+        replaceContainsTypeFlag(*parentframe, ChildType::Delete);
+      else
+        deleteOneNode(node);
+    }
+  }
+}
+
+void removeFromTreeIterative(execplan::ParseTree** root, const CommonContainer& common)
+{
+  if (common.first.empty())
+    return;
+
+  DFSStack stack;
+  stack.emplace_back(root, GoTo::Left);
+  while (!stack.empty())
+  {
+    auto [node, flag, ltype, rtype] = stack.back();
+    if (flag != GoTo::Up)
+    {
+      addStackFrame(stack, flag, *node);
+      continue;
+    }
+
+    auto sz = stack.size();
+    if (castToFilter(*node) && sz > 1)
+    {
+      if (commonContainsPtr(common, *node))
+        replaceContainsTypeFlag(stack.at(sz - 2), ChildType::Unchain);
+      else if (!commonContainsPtr(common, *node) && commonContainsSemantic(common, *node))
+        replaceContainsTypeFlag(stack.at(sz - 2), ChildType::Delete);
+      else
+        replaceContainsTypeFlag(stack.at(sz - 2), ChildType::Leave);
+
+      stack.pop_back();
+      continue;
+    }
+
+    if (sz == 1)
+      fixUpTree(node, ltype, rtype);
+    else
+      fixUpTree(node, ltype, rtype, &stack[sz - 2]);
+
+    stack.pop_back();
+  }
+}
+
+}  // namespace details
+
+void dumpTreeFiles(execplan::ParseTree* filters, const std::string& name, std::string dumpfolder = {})
+{
+#if debug_rewrites
+  messageqcpp::ByteStream beforetree;
+  ObjectReader::writeParseTree(filters, beforetree);
+
+  if (dumpfolder.empty())
+  {
+    dumpfolder = startup::StartUp::tmpDir();
+  }
+
+  std::ofstream before(dumpfolder + "filters." + name + ".data");
+  before << beforetree;
+  std::string dotname = dumpfolder + "filters." + name + ".dot";
+  filters->drawTree(dotname);
+  std::string dotInvoke = "dot -Tpng ";
+  std::string convert = dotInvoke + dotname + " -o " + dotname + ".png";
+  [[maybe_unused]] auto _ = std::system(convert.c_str());
+#endif
+}
+
+template <bool stableSort>
+execplan::ParseTree* extractCommonLeafConjunctionsToRoot(execplan::ParseTree* tree)
```

Comment:
Some thoughts for the future - what do you think about unification for pattern match/rewrite, for example:
1. We have unified tree traversal.
2. We have a list of patterns - pattern 1, pattern 2, pattern 3, they are applied iteratively and after each rewrites happens.
If we have some unified infrastructure we do not need to write again tree traversal, we can just specify what to do on some callbacks.
BTW, my point is not about to rewrite this current solution, I'm just interesting what do you think in general, may be it's not applicable here.

Reply:
I like the idea and in fact I was thinking about something like that. However I have to say that the code here is somewhat specific for one particular case -- for example the collectCommonConjunctions function wouldn't work for the purpose of this transformation `a or (a and b) => a`. And it's not obvious to me how to change it right now so it would work.

I had a completely different approach in mind initially, which might've been more generalizeable. Maybe I'll come back to it at some point.